### ISSUE TITLE
The form of loss functions

### ISSUE THREAD

**Issue by ablaom** [alan-turing-institute/MLJ.jl#91]
At present we are developing our own loss function code, partly because of the needs to address probabilistic predictors, and partly because of the relatively immature, unmaintained state of MLMetrics. (I suppose that ultimately MLJ may drive a revival of that package.)

For now, our code for dealing with metrics takes a metric to be of the form `some_metric(y, yhat)`, where `y` is a vector of true target values and `yhat`, the predicted values, or predicted probability distributions. Two issues to be nailed down are:

(i) how to deal with multivariate targets. Presumably, we simply allow arrays in addition to vector inputs, and give the user a way to construct "product" metrics, no?

(ii) how to deal with losses that depend not just on the target but on inputs. For example, the loss function for a supermarket sales prediction might give more weight to perishable items. One solution might be to bundle the extra information with the target (eg, call the perishable flag a second target whose prediction is trivial) but is this solution ideal? Another solution (sort-of adopted in Flux but for possibly different reasons) is to let metrics have the form `some_metric(y, X, machine)` and get the metric to call the `predict` method on `X` to get `yhat`. Thoughts?

**Comment by fkiraly** [comment]
(i) that would be my guess too. The "product" metric, in the language of losses, is simply the sum of losses, over prediction components.
A slight complication arises when the prediction is probabilistic, but not itself a product (which is the product distribution in the "category" of probabilistic prediction). In that case, the product loss would have to act on the marginals, which is maybe best avoided since this is not explicit and would complicate the interface.
Therefore, I do not think that standard behaviour should be the product - the user should construct multivariate losses/scores themselves, and there should be standards which "know" that they are multivariate.

(ii) I'd simply give them the form ``loss(yhat,y,w)`` where w is a vector or data frame of equal length as y and yhat.

Further unsolicited comments: 
(iii) as a matter of taste, I'd put yhat *first*, i.e., ``metric(yhat, y)``, and call it score, or loss, and not metric. I've seen this causing user confusion before where the user looks up "metric" on the internet and thinks it is this: https://en.wikipedia.org/wiki/Metric_(mathematics)

(iv) a point to consider is that a number of scores are computed efficiently when computed together - this situation may arise frequently when a user wishes to compute very similar scores, e.g., RMSE and MAE and R-squared. In some (unpublished) instances before I've bundled these together in scorer objects that compute multiple scores together, at nil additional cost, but return only those queried by the user.

**Comment by fkiraly** [comment]
And (v), I think scorers/losses should be proper structs, potentially indexed by the tasks they are applicable to. E.g., something more along the lines of ``score(rmse, yhat, y)`` maybe?

**Comment by kirtsar** [comment]
I believe you don't have to reinvent the wheel.
Loss functions package is pretty mature : https://github.com/JuliaML/LossFunctions.jl
There are bunch  of different metrics with nice API

**Comment by fkiraly** [comment]
Oooh, thanks, @kirtsar , this looks pretty nice - and it follows my API suggestions, including dispatch and element-wise computations (except for having the predictions last not first).

It should be easy to extend this to probabilistic losses, too - but that is not implemented yet as far as I can see.

**Comment by kirtsar** [comment]
Also for algorithms based on metrics you should definitely know about this package:
https://github.com/JuliaStats/Distances.jl

**Comment by fkiraly** [comment]
Regarding LossFunctions.jl, see discussion there on probabilistic losses:
https://github.com/JuliaML/LossFunctions.jl/issues/108

What I take from this is that it is no longer in active development (?), thus it seems unlikely that the devs would add functionality for probabilistic losses.
While a few probabilistic losses are implemented for classification, it is in the sklearn style interface which we had decided against a while ago
https://github.com/alan-turing-institute/MLJ.jl/issues/34
(because it messes up the probabilistic classification API)

Though, what we could do, since all of this is open source, is copy all the nice parts and add/remove as we see fit?

Still think the predictions should be the first argument, due to the logical dispatch order - e.g., in the example of the log-loss ``logloss(y_pred, y) = - log ( pdf( y_pred, y) )``, methods of the prediction distribution are called with the observations as an argument, making the (prediction) distribution the natural first argument of the dispatch.a

**Comment by ValdarT** [comment]
> Though, what we could do, since all of this is open source, is copy all the nice parts and add/remove as we see fit?

Why not contribute to the `LossFunctions.jl` directly? Even if it's not in active development, they would probably be happy to accept and review PR-s.

**Comment by fkiraly** [comment]
@ValdarT , obviously also sth to consider, but there is a clash when it comes to probabilistic classification losses. We'd need it to dispatch on distributions, they dispatch on probabilities.

So this would involve at least an extension (or rewrite) of interface which may or may not fit with the LossFunctions.jl vision.

**Comment by kirtsar** [comment]
I believe the extension is pretty straightforward. Given the loss function, if one of its argument is Distribution, then do this, else call loss directly:

```
function (foo :: T)(d :: D, y) where T <: SupervisedLoss, D <: Distribution
return foo(pdf(d, y))
```

**Comment by fkiraly** [comment]
@kirtsar I would actually do it the other way round - if a number is passed, make it into a binomial probability distribution.

This is because: 
- it avoids the hassle with "how to represent multi-class probabilistic predictions", since it uses a common interface.
- not every loss is a function of the pdf/pmf evaluate. In fact, the only strictly proper loss of this form is the log-loss (= cross-entropy loss). An example which you couldn't cover in the above way is the Brier score, which is
``2*pmf(d,y) - l2norm(d)``, and the continuous variant is ``2*pdf(d,y) - l2norm(d)``

**Comment by fkiraly** [comment]
Random side remark: poisson loss and log-loss (aka cross-entropy loss) are *the same probabilistic loss*, just using different (log)pmf. This is, i.m.o. an instance of interface proliferation due to a sub-optimally chosen design...

**Comment by ablaom** [comment]
This all sounds like good. All we need is a volunteer to implement :-) 

BTW, in upcoming push to master, the order of `y` and `ypred` will be reversed (`ypred` first).

**Comment by fkiraly** [comment]
So, @blaom, regarding the volunteering, there are maybe a few technical questions to clarify ahead:
(i) where should it go? LossFunctions.jl, and MLJ.jl, and within the projects where (neither having the full machinery at the moment)
(ii) how to deal with "vectors of distributions" precisely. Have we agreed on a type here? 
(iii) similarly, how to deal with a completely non-parametric predictive distributions which may have to check/access the fitted model rather than explicit parameters
(iv) what do we do about the missing methods in Distributions.jl, e.g., l2-norm-of-pdf

**Comment by ablaom** [comment]
Yes, the Distributions.Distribution type is too restrictive to deal with mixed-type multivariate targets, non-parametric distributions (eg, sampling by MCMC), non-linear supports (eg, spheres in geophysics), and so forth. I think we are more-or-less forced here to introduce our more general concept of a distribution as *trait*, rather than a type, in analogy with the `istable` trait of Tables.jl.  I wonder if something similar has not already been begun elsewhere?

**Comment by datnamer** [comment]
perhaps th Turing.jl teamm might have some ideas on this (or integration?)

cc @yebai @mohamed82008 @trappmartin @xukai92

**Comment by mohamed82008** [comment]
Thanks for pinging me, this looks like a great package! I read the discussion above but I probably lack some context so my comments may be out of place. Here are some thoughts anyways:
1. I don't see how `Distributions.Distribution` is restricting. You can define a new `MixedContinuousDiscrete <: ValueSupport` for mixed continuous and discrete multivariate distributions, and then subtype `Distribution{Multivariate, MixedContinuousDiscrete}`.
2. Subtyping `Distribution` is one path. Another is to simply assume an interface is defined on the new `Distribution`-like type and leave that argument unrestricted, `::Any`, in `MLJ` functions. So multiple dispatch can happen at a lower level when the specialized methods are called. The interface can include `rand(dist_like_struct)` for example and its variants or any other useful method. A trait can be used to formalize this interface or for more complex behavior specializations, but maybe overkill. `Turing.Sampler` will probably end up as a `Distribution`-like struct that defines `Distribution`-esque methods without subtyping `Distribution`.
3. If there are specific requirements from the `Turing.jl` side to make it compatible with `MLJ`, please open an issue in `Turing.jl` with a piece of code that you hope to see working but currently isn't. There are significant redesign PRs and issues currently open and being addressed on the Turing side so this is a great time to make your wish list known to core devs.

**Comment by ablaom** [comment]
Yes, I see one can also just subtype Distributions.Sampleable for things that just implement `rand`.

Currently MLJ uses the custom `UnivariateFinite` for distributions with a support that is either a set of `CategoricalValue` or `CategoricalString` elements. I couldn't see an elegant way to use the existing distributions whose supports are in `Int` or `Float64`. I suppose we could add `Categorical` to the existing subtypes of `Discrete`, `Continuous` of `ValueSupport` and type `UnivariateFinite` as a subtype of `Distributions{Univariate,Categorical}`.

**Comment by fkiraly** [comment]
@ablaom, counts, or integers (incl negative) are an important but not-so-frequently-used class of distributions.
There are also specific models, loss functions, and pipeline composition techniques for counts or integers - so I would suggest to mirror this both on the distribution type side as well as on the model side.
For example, where should GLM with Poisson link, or a neural network with Poisson output layer sit?

There is also an important distinction to make between:
- distributions that take values in a finite number of categories
- distributions that take values in a finite (or countable) sub-set of the reals

The second can arise as full distributional predictions - e.g., posterior samples from MCMC style Bayesian methods - rather than always being outputs of a probabilistic classifier. Main difference is the finite support that may vary across a potentially uncountably infinite set (or not).

**Comment by ablaom** [comment]
For tuning (and evaluating methods called by tuners) it would be nice if loss functions had traits, with `true` the fallback value:

- iscustom (true for user-defined losses, that are actually functions)

- isloss (true if we are minimising, false if we are maximising)

- isanaggregate (rms is an aggregate but AUC is not)

**Comment by ablaom** [comment]
[Update](https://github.com/alan-turing-institute/MLJ.jl/blob/master/docs/src/measures.md) in the manual showing current state of the API.

**Comment by ablaom** [comment]
Thanks to offline discussions with @fkiraly, I think we now have a reasonable flushed out loss functions API in place, as documented in the manual. Still to do: 

- product loss functions for multi-targets (looks pretty straight-forward),

- implementation of proper scoring rules (with necessary supporting PR's to Distributions)

- more thought perhaps around probabilistic predictions that are only sampleable (no pdf/mdf)

**Comment by ablaom** [comment]
This has now been implemented.
